File: marketdna-backend/app/models/indicators.py
"""Pydantic models for Technical Indicator Analysis."""
from pydantic import BaseModel


class TrendReading(BaseModel):
    sma20: float
    sma50: float
    sma100: float
    sma200: float
    ema20: float
    ema50: float
    price_vs_sma20: float
    price_vs_sma50: float
    price_vs_sma100: float
    price_vs_sma200: float
    alignment_score: int        # 0–4 (count of SMAs price is above)
    trend_label: str            # "Fully Bullish" | "Mostly Bullish" | "Mixed" | "Mostly Bearish" | "Fully Bearish"


class MomentumReading(BaseModel):
    rsi14: float
    rsi_signal: str             # "Oversold" | "Near Oversold" | "Neutral" | "Near Overbought" | "Overbought"
    macd_line: float
    macd_signal: float
    macd_hist: float
    macd_crossover: str         # "Bullish" | "Bearish" | "None"
    stoch_k: float
    stoch_d: float
    stoch_signal: str           # "Overbought" | "Oversold" | "Neutral"


class VolatilityReading(BaseModel):
    bb_upper: float
    bb_middle: float
    bb_lower: float
    bb_pct_b: float             # 0–100; 0 = at/below lower band, 100 = at/above upper band
    bb_signal: str
    atr14: float
    atr_pct: float              # ATR as % of price


class VolumeReading(BaseModel):
    obv: float
    volume: float
    volume_sma20: float
    volume_ratio: float         # current / 20d avg
    volume_signal: str          # "High Volume" | "Low Volume" | "Normal"


class IndicatorDivergence(BaseModel):
    type: str
    description: str


class StockIndicators(BaseModel):
    symbol: str
    date: str
    price: float
    trend: TrendReading
    momentum: MomentumReading
    volatility: VolatilityReading
    volume: VolumeReading
    divergences: list[IndicatorDivergence]
    overall_signal: str         # "Strong Buy" | "Buy" | "Neutral" | "Sell" | "Strong Sell"
    signal_score: float
    computed_at: str


class IndicatorScanItem(BaseModel):
    symbol: str
    price: float
    rsi14: float
    rsi_signal: str
    macd_hist: float
    macd_crossover: str
    bb_pct_b: float
    bb_signal: str
    atr_pct: float
    volume_ratio: float
    trend_label: str
    alignment_score: int
    overall_signal: str
    signal_score: float
    signals: list[str]


class IndicatorScanResult(BaseModel):
    items: list[IndicatorScanItem]
    signal_distribution: dict[str, int]
    computed_at: str

File: marketdna-web/src/types/indicators.ts
export const SIGNAL_LABELS = ["Strong Buy", "Buy", "Neutral", "Sell", "Strong Sell"] as const
export type SignalLabel = typeof SIGNAL_LABELS[number]

export const SIGNAL_COLOR: Record<string, string> = {
  "Strong Buy":  "#22c55e",
  "Buy":         "#86efac",
  "Neutral":     "#94a3b8",
  "Sell":        "#f87171",
  "Strong Sell": "#ef4444",
}

export const RSI_SIGNAL_COLOR: Record<string, string> = {
  "Oversold":        "#22c55e",
  "Near Oversold":   "#86efac",
  "Neutral":         "#94a3b8",
  "Near Overbought": "#fbbf24",
  "Overbought":      "#ef4444",
}

export const TREND_COLOR: Record<string, string> = {
  "Fully Bullish":   "#22c55e",
  "Mostly Bullish":  "#86efac",
  "Mixed":           "#94a3b8",
  "Mostly Bearish":  "#f87171",
  "Fully Bearish":   "#ef4444",
}

export const BB_SIGNAL_COLOR: Record<string, string> = {
  "Below Lower Band": "#22c55e",
  "Near Lower Band":  "#86efac",
  "Mid Band":         "#94a3b8",
  "Near Upper Band":  "#fbbf24",
  "Above Upper Band": "#ef4444",
}

export const MACD_COLOR: Record<string, string> = {
  "Bullish": "#22c55e",
  "Bearish": "#ef4444",
  "None":    "#94a3b8",
}

export interface TrendReading {
  sma20: number
  sma50: number
  sma100: number
  sma200: number
  ema20: number
  ema50: number
  price_vs_sma20: number
  price_vs_sma50: number
  price_vs_sma100: number
  price_vs_sma200: number
  alignment_score: number
  trend_label: string
}

export interface MomentumReading {
  rsi14: number
  rsi_signal: string
  macd_line: number
  macd_signal: number
  macd_hist: number
  macd_crossover: string
  stoch_k: number
  stoch_d: number
  stoch_signal: string
}

export interface VolatilityReading {
  bb_upper: number
  bb_middle: number
  bb_lower: number
  bb_pct_b: number
  bb_signal: string
  atr14: number
  atr_pct: number
}

export interface VolumeReading {
  obv: number
  volume: number
  volume_sma20: number
  volume_ratio: number
  volume_signal: string
}

export interface IndicatorDivergence {
  type: string
  description: string
}

export interface StockIndicators {
  symbol: string
  date: string
  price: number
  trend: TrendReading
  momentum: MomentumReading
  volatility: VolatilityReading
  volume: VolumeReading
  divergences: IndicatorDivergence[]
  overall_signal: string
  signal_score: number
  computed_at: string
}

export interface IndicatorScanItem {
  symbol: string
  price: number
  rsi14: number
  rsi_signal: string
  macd_hist: number
  macd_crossover: string
  bb_pct_b: number
  bb_signal: string
  atr_pct: number
  volume_ratio: number
  trend_label: string
  alignment_score: number
  overall_signal: string
  signal_score: number
  signals: string[]
}

export interface IndicatorScanResult {
  items: IndicatorScanItem[]
  signal_distribution: Record<string, number>
  computed_at: string
}
